# Homework 1

> [!NOTE]
> This homework uses the pinned 2026 car fuel-efficiency release in the course
> repository. The plan and report are available in `cohorts/2026/data/`.

## Set up the environment

You need to install Python, NumPy, Pandas, Matplotlib, and Seaborn. For that, you can use the instructions from
[06-environment.md](06-environment.md).

## Q1. Pandas version

What version of Pandas did you install?

You can get the version information using the `__version__` field:

```python
pd.__version__
```

In [6]:
import numpy as np
import pandas as pd

In [7]:
pd.__version__

'3.0.3'


## Getting the data 

For this homework, we'll use the 2026 Car Fuel Efficiency dataset. Download it from <a href='https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/main/cohorts/2026/data/car_fuel_efficiency_2026.csv'>here</a>.

You can do it with wget:
```bash
wget https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/main/cohorts/2026/data/car_fuel_efficiency_2026.csv
```

Or just open it with your browser and click "Save as...".

Now read it with Pandas.

In [9]:
df = pd.read_csv('./car_fuel_efficiency_2026.csv')

In [10]:
df.head()

,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
0,2006,Europe,Gasoline,Front-wheel drive,4,2180,6,243.0,3870,NaN,31.9
1,2008,Europe,Diesel,Front-wheel drive,4,2390,6,272.0,4210,NaN,31.3
2,1996,Asia,Gasoline,Front-wheel drive,5,2320,6,267.0,4240,17.3,27.5
3,1989,Europe,Gasoline,Front-wheel drive,4,2130,6,258.0,4490,18.7,28.5
4,1994,USA,Diesel,Front-wheel drive,3,2580,7,304.0,4510,17.5,31.0


## Q2. Records count

How many records are in the dataset?

- 5000
- 9000
- 10000
- 15000

In [12]:
df.describe()

,model_year,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
count,10000.000000,10000.00000,10000.000000,10000.000000,9123.000000,10000.000000,9736.000000,10000.000000
mean,1999.549500,3.50400,2268.807000,6.035200,254.446016,4286.754000,18.888044,29.997700
std,14.207925,0.95209,183.447355,0.542578,22.844613,266.212491,1.279543,2.930245
min,1975.000000,2.00000,1590.000000,4.000000,171.000000,3320.000000,14.000000,19.800000
25%,1987.000000,3.00000,2150.000000,6.000000,239.000000,4110.000000,18.000000,28.000000
50%,2000.000000,4.00000,2270.000000,6.000000,254.000000,4280.000000,18.900000,30.000000
75%,2012.000000,4.00000,2390.000000,6.000000,270.000000,4470.000000,19.700000,31.900000
max,2024.000000,5.00000,3110.000000,8.000000,334.000000,5460.000000,23.800000,41.200000


In [13]:
df.shape

(10000, 11)

The answer is **10000**.

## Q3. Fuel types

How many fuel types are presented in the dataset?

- 1
- 2
- 3
- 4

In [14]:
df.fuel_type.describe()

count        10000
unique           3
top       Gasoline
freq          6913
Name: fuel_type, dtype: object

The answer is **3**.


## Q4. Missing values

How many columns in the dataset have missing values?

- 0
- 1
- 2
- 3
- 4

In [16]:
df.isna().sum()

model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower             877
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64

The answer is **2**.

## Q5. Max fuel efficiency

What's the maximum fuel efficiency of cars from Asia?

- 21.2
- 31.2
- 41.2
- 51.2

In [21]:
df[df['origin'] == 'Asia']['fuel_efficiency_mpg'].max()

np.float64(41.2)

The answer is **41.2**.

## Q6. Median value of horsepower

1. Find the median value of the `horsepower` column in the dataset.
2. Next, calculate the most frequent value of the same `horsepower` column.
3. Use the `fillna` method to fill the missing values in the `horsepower` column with the most frequent value from the previous step.
4. Now, calculate the median value of `horsepower` once again.

Has it changed?


- Yes, it increased
- Yes, it decreased
- No

In [27]:
df.horsepower.median()

np.float64(254.0)

In [28]:
df.horsepower.value_counts()

horsepower
252.0    174
246.0    168
256.0    168
259.0    165
253.0    165
        ... 
322.0      1
330.0      1
175.0      1
178.0      1
320.0      1
Name: count, Length: 153, dtype: int64

In [31]:
df.horsepower = df.horsepower.fillna(252)

In [32]:
df.horsepower.median()

np.float64(252.0)

The median **decreased**.

## Q7. Sum of weights

1. Select all the cars from Asia
2. Select only columns `vehicle_weight` and `model_year`
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it `X`.
5. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
6. Invert `XTX`.
7. Create an array `y` with values `[1100, 1300, 800, 900, 1000, 1100, 1200]`.
8. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
9. What's the sum of all the elements of the result?

> **Note**: You just implemented linear regression. We'll talk about it in the next lesson.

- 0.0369
- 0.369
- 3.69
- 36.9

## Learning in Public

As part of this course, we encourage you to learn in public. Sharing your progress helps you reflect, stay accountable, and connect with others.

Write posts on your social media about Module 1 of this course.

How to do it
- Pick one platform (or more!) and post using the templates below
- Don't forget to tag Alexey Grigorev and use #mlzoomcamp hashtag

Alexey on social media: [LinkedIn](https://www.linkedin.com/in/agrigorev/), [Twitter/X](https://x.com/Al_Grigor)
DTC on social media: [LinkedIn](https://www.linkedin.com/company/datatalks-club/), [Twitter/X](https://x.com/DataTalksClub)

---

### LinkedIn Template

I just completed Module 1: Introduction to Machine Learning from ML Zoomcamp 2026 🎉

This module covered:

🔹 Difference between ML and rule-based systems

🔹 What supervised machine learning is

🔹 CRISP-DM: a framework for structuring ML projects

🔹 The model selection step

🔹 Setting up the environment (Python, Jupyter, etc.)

🔹 Quick refreshers on NumPy, linear algebra, and Pandas

✨ My key takeaway: [insert your personal insight]

🤔 Something I struggled with or found interesting: [insert challenge, surprise, or aha-moment]

➡️ Next step → Module 2: [your personal goal or excitement]

#mlzoomcamp @Alexey Grigorev

---

### Twitter / X Template

#### Single tweet version
Just wrapped up Module 1 of #mlzoomcamp 🚀

Learned about:

🔹 ML vs. rule-based systems

🔹 Supervised ML basics

🔹 CRISP-DM framework

🔹 Setting up environment + NumPy/Pandas/Linear Algebra refreshers

✨ Key takeaway: [insert insight]

➡️ Next up: Module 2!

@Al_Grigor

#### Thread starter
Day 1 of #mlzoomcamp ✅

Module 1: Intro to ML

Here’s what I learned 👇

(then add 2–3 tweets: 💡 takeaway, 🤔 challenge, 🚀 goal for Module 2)


## Submit the results

* Submit your results here: https://courses.datatalks.club/ml-zoomcamp-2026/homework/hw01
* The numerical options are calculated from the pinned 2026 release. Use the value that matches your calculation.